### PR TITLE
Fix eclipse issues related to rest client shading

### PR DESCRIPTION
**Pull request by hub-cap**
* A cycle was detected in eclipse, and was fixed in the same fashion as
  core and core-tests.
* The rest client deps jar was not properly exported in the generated
  eclipse classpath file for rest client.
* Moved the plugin dependency due to eclipse-build 'apply from' not
  liking the plugin declaration in the rest client build.

Relates #25208

### CODE REVIEW

**Comment by hub-cap** [comment]
@nik9000 just an fyi, there is still a single problem in the eclipse build, but its totally unrelated to (or at least I think) the changes in the rest client.

File: buildSrc/build.gradle [hunk 1]
```diff
@@ -94,6 +94,7 @@ dependencies {
   compile 'com.perforce:p4java:2012.3.551082' // THIS IS SUPPOSED TO BE OPTIONAL IN THE FUTURE....
   compile 'de.thetaphi:forbiddenapis:2.3'
   compile 'org.apache.rat:apache-rat:0.11'
+  compile 'com.github.jengelman.gradle.plugins:shadow:2.0.1'
```
Comment (rjernst) [inline buildSrc/build.gradle]: Why is this needed? Nothing in buildSrc uses it right?

File: client/rest/build.gradle [hunk 2]
```diff
@@ -39,11 +39,8 @@ import org.elasticsearch.gradle.precommit.PrecommitTasks
  * 3) The *actual* jar that will be used by clients. This has no classifier, contains the rest client src and
  *    `org.elasticsearch.client`. This jar is the only actual output artifact of this job.
  */
-plugins {
-  id "com.github.johnrengelman.shadow" version "2.0.1"
```
Comment (rjernst) [inline client/rest/build.gradle]: Why are you moving this to the "apply plugin" form?

**Review by rjernst — APPROVED** [review]
LGTM, just a question on why we need the shadow plugin in buildSrc

**Comment by colings86** [comment]
Just for information, I have tested this on "Eclipse Neon.3 Release (4.6.3)" and can confirm it fixes the dependency cycle leaving no issues in Eclipse

**Comment by colings86** [comment]
I also tested master on "Eclipse Oxygen Release (4.7.0)" and the dependency cycle issue is fixed but I get the following error:
```
Description	Resource	Path	Location	Type
Cannot infer type argument(s) for <T> checkEqualsAndHashCode(T, EqualsHashCodeTestUtils.CopyFunction<T>, EqualsHashCodeTestUtils.MutateFunction<T>)	SuggestBuilderTests.java	/:core-tests/java/org/elasticsearch/search/suggest	line 90
```

I assume this is the error that @hub-cap is referring to above. These kinds of generics issues are common in eclipse unfortunately and seems unrelated to the issue this PR is fixing

**Comment by nik9000** [comment]
> These kinds of generics issues are common in eclipse unfortunately and seems unrelated to the issue this PR is fixing

We tend to push Eclipse's compiler quite a bit, yeah. I've been filing issues with them as we encounter the issues and working around them and leaving a comment linking to the issue.

File: client/rest/build.gradle [hunk 3]
```diff
@@ -109,6 +106,18 @@ if (isIdea == false && isEclipse == false) {
   }
 }
 
+if (isEclipse) {
+  // in eclipse the project is under a fake root, we need to change around the source sets
+  sourceSets {
+    if (project.path == ":client:rest") {
+      main.java.srcDirs = ['java']
+      //main.resources.srcDirs = ['resources']
```
Comment (nik9000) [inline client/rest/build.gradle]: I'd either uncomment or remove this line entirely.

**Comment by nik9000** [comment]
I tested locally and this fixes my issue.

**Review by nik9000 — APPROVED** [review]

**Comment by hub-cap** [comment]
I thought I was crazy for a moment, but this indeed (with that reverted second commit) blows up when doing anything eclipse related. I will attempt to use the buildscript style.

https://docs.gradle.org/3.3/userguide/plugins.html#plugins_dsl_limitations

```
The plugins {} block can currently only be used in a project's build script. It cannot be used in script plugins, the settings.gradle file or init scripts.

Future versions of Gradle will remove this restriction.

If the restrictions of the new syntax are prohibitive, the recommended approach is to apply plugins using the buildscript {} block.
```

**Comment by hub-cap** [comment]
2c271f0f222b47ecba4aa68c887eb5f5e36f5050 fixes the last comment I made.
